File: level1/lib/core/detector.dart
import '../core/models.dart';
import '../detection/audio/audio_classifier.dart';

class DetectorConfig {
  final String type;
  final bool enabled;
  final double threshold;
  final int persistenceFrames;
  final Duration cooldown;
  final List<String> routeToChannelIds;
  final bool motionGated;

  const DetectorConfig({
    required this.type,
    this.enabled = true,
    this.threshold = 0.5,
    this.persistenceFrames = 2,
    this.cooldown = const Duration(seconds: 60),
    this.routeToChannelIds = const [],
    this.motionGated = false,
  });

  DetectorConfig copyWith({
    String? type,
    bool? enabled,
    double? threshold,
    int? persistenceFrames,
    Duration? cooldown,
    List<String>? routeToChannelIds,
    bool? motionGated,
  }) {
    return DetectorConfig(
      type: type ?? this.type,
      enabled: enabled ?? this.enabled,
      threshold: threshold ?? this.threshold,
      persistenceFrames: persistenceFrames ?? this.persistenceFrames,
      cooldown: cooldown ?? this.cooldown,
      routeToChannelIds: routeToChannelIds ?? this.routeToChannelIds,
      motionGated: motionGated ?? this.motionGated,
    );
  }

  Map<String, dynamic> toJson() => {
        'type': type,
        'enabled': enabled,
        'threshold': threshold,
        'persistenceFrames': persistenceFrames,
        'cooldownMs': cooldown.inMilliseconds,
        'routeToChannelIds': routeToChannelIds,
        'motionGated': motionGated,
      };

  factory DetectorConfig.fromJson(Map<String, dynamic> json) => DetectorConfig(
        type: json['type'] as String,
        enabled: json['enabled'] as bool? ?? true,
        threshold: (json['threshold'] as num?)?.toDouble() ?? 0.5,
        persistenceFrames: json['persistenceFrames'] as int? ?? 2,
        cooldown: Duration(milliseconds: json['cooldownMs'] as int? ?? 60000),
        routeToChannelIds:
            (json['routeToChannelIds'] as List?)?.cast<String>() ?? const [],
        motionGated: json['motionGated'] as bool? ?? false,
      );
}

abstract class Detector {
  String get id;

  DetectorConfig get config;

  String get triggerType;

  Future<void> init();

  void reset();

  Future<void> dispose();
}

abstract class FrameDetector extends Detector {
  /// Inclusion regions (normalized 0..1 on the analysis frame). Empty = detect
  /// everywhere. Set by the pipeline via [DetectorPipeline.setRegions].
  List<DetectionRegion> regions = const [];

  DetectionResult analyzeFrame(AnalysisFrame frame);

  /// Async analysis path for gated/heavy detectors (runs off the pipeline's
  /// sync per-frame loop). Defaults to the sync path wrapped in a Future.
  Future<DetectionResult> analyzeFrameAsync(AnalysisFrame frame) async =>
      analyzeFrame(frame);
}

abstract class AudioDetector extends Detector {
  DetectionResult analyzeScores(AudioEventScores scores);
}

File: level1/android/app/src/main/kotlin/io/securitycam/level1/detection/Detector.kt
package io.securitycam.level1.detection

import io.securitycam.level1.detection.audio.AudioEventScores
import java.time.Duration

/**
 * Detector configuration (port of `lib/core/detector.dart` `DetectorConfig`).
 * JSON keys match the Dart blob.
 */
data class DetectorConfig(
    val type: String,
    val enabled: Boolean = true,
    val threshold: Double = 0.5,
    val persistenceFrames: Int = 2,
    val cooldown: Duration = Duration.ofSeconds(60),
    val routeToChannelIds: List<String> = emptyList(),
    val motionGated: Boolean = false,
) {
    fun copyWith(
        type: String? = null,
        enabled: Boolean? = null,
        threshold: Double? = null,
        persistenceFrames: Int? = null,
        cooldown: Duration? = null,
        routeToChannelIds: List<String>? = null,
        motionGated: Boolean? = null,
    ): DetectorConfig = DetectorConfig(
        type = type ?: this.type,
        enabled = enabled ?: this.enabled,
        threshold = threshold ?: this.threshold,
        persistenceFrames = persistenceFrames ?: this.persistenceFrames,
        cooldown = cooldown ?: this.cooldown,
        routeToChannelIds = routeToChannelIds ?: this.routeToChannelIds,
        motionGated = motionGated ?: this.motionGated,
    )

    fun toJson(): Map<String, Any?> = mapOf(
        "type" to type,
        "enabled" to enabled,
        "threshold" to threshold,
        "persistenceFrames" to persistenceFrames,
        "cooldownMs" to cooldown.toMillis(),
        "routeToChannelIds" to routeToChannelIds,
        "motionGated" to motionGated,
    )

    companion object {
        fun fromJson(json: Map<String, Any?>): DetectorConfig = DetectorConfig(
            type = json["type"] as String,
            enabled = json["enabled"] as? Boolean ?: true,
            threshold = (json["threshold"] as? Number)?.toDouble() ?: 0.5,
            persistenceFrames = (json["persistenceFrames"] as? Number)?.toInt() ?: 2,
            cooldown = Duration.ofMillis(
                (json["cooldownMs"] as? Number)?.toLong() ?: 60_000L,
            ),
            routeToChannelIds = (json["routeToChannelIds"] as? List<*>)
                ?.map { it as String }
                ?: emptyList(),
            motionGated = json["motionGated"] as? Boolean ?: false,
        )
    }
}

/** Base detector contract (port of `lib/core/detector.dart`). */
interface Detector {
    val id: String
    val config: DetectorConfig
    val triggerType: String

    suspend fun init()

    fun reset()

    suspend fun dispose()
}

/** Frame-based detector; [regions]/[exclusionRegions] are set by the pipeline. */
abstract class FrameDetector : Detector {
    var regions: List<DetectionRegion> = emptyList()
    var exclusionRegions: List<DetectionRegion> = emptyList()

    abstract fun analyzeFrame(frame: AnalysisFrame): DetectionResult

    /** Async analysis path for gated/heavy detectors; defaults to the sync path. */
    open suspend fun analyzeFrameAsync(frame: AnalysisFrame): DetectionResult =
        analyzeFrame(frame)
}

/** Score-based audio detector. */
abstract class AudioDetector : Detector {
    abstract fun analyzeScores(scores: AudioEventScores): DetectionResult
}